### ISSUE TITLE
In Unix is not found with a dot end path dll file

### ISSUE THREAD

**Issue by hueifeng** [dotnet/runtime#37112]
# Steps to Reproduce
```
ctrl-macbookpro2159@CTRL-MacBook-Pro test. % dotnet run 
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly '/Users/test./bin/Debug/netcoreapp3.1/test..dll'. The system cannot find the file specified.

File name: '/Users/test./bin/Debug/netcoreapp3.1/test..dll'


ctrl-macbookpro2159@CTRL-MacBook-Pro test. % cd /Users/test 
ctrl-macbookpro2159@CTRL-MacBook-Pro test % dotnet run
Hello World!
ctrl-macbookpro2159@CTRL-MacBook-Pro test %
```

# OS & dotnet version

OS Name: MacOs Catalina
OS Version: 10.15.3
OS Platform: Darwin

dotnet SDK:  3.1.300

**Comment by danmosemsft** [comment]
Does COREHOST_TRACE give any insight?
https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing#how-do-i-debug-the-probing-properties-construction

**Comment by hueifeng** [comment]
@danmosemsft Running in macOS, the effect is as follows. But I can print the information by setting the environment variable by running the same command under the windows system.
```
ctrl-macbookpro2159@CTRL-MacBook-Pro test. % set COREHOST_TRACE=1
ctrl-macbookpro2159@CTRL-MacBook-Pro test. % dotnet run          
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly '/Users/test./bin/Debug/netcoreapp3.1/test..dll'. The system cannot find the file specified.
File name: '/Users/test./bin/Debug/netcoreapp3.1/test..dll'
ctrl-macbookpro2159@CTRL-MacBook-Pro test. %
```
The result of the path without dots is shown below:
```
ctrl-macbookpro2159@CTRL-MacBook-Pro test % set COREHOST_TRACE=1
ctrl-macbookpro2159@CTRL-MacBook-Pro test % dotnet run          
Hello World!
ctrl-macbookpro2159@CTRL-MacBook-Pro test %
```

**Comment by danmosemsft** [comment]
I think you have to set COREHOST_TRACEFILE=<path> and then look in that file

**Comment by hueifeng** [comment]
@danmosemsft   I do n’t know why I executed the command on macOS, but still, no result shows me.

**Comment by danmosemsft** [comment]
Did it even write the file you specified?

**Comment by hueifeng** [comment]
@danmosemsft I did this by executing the command to set up the environment, but did not automatically create and write files

**Comment by danmosemsft** [comment]
What path did you set COREHOST_TRACEFILE to?

**Comment by hueifeng** [comment]
@danmosemsft  I tried the following four commands, but none of them displayed as expected.
```
set COREHOST_TRACEFILE=text.txt
set COREHOST_TRACEFILE=text
set COREHOST_TRACEFILE=/Users/text.txt
set COREHOST_TRACEFILE=/Users/text
```

**Comment by danmosemsft** [comment]
And also set COREHOST_TRACE=1

**Comment by hueifeng** [comment]
Currently I don't use the folder suffix path ending in ` . ` . The execution results are as follows:
```
Last login: Thu May 28 11:25:09 on ttys005
ctrl-macbookpro2159@CTRL-MacBook-Pro ~ % cd /Users/test 
ctrl-macbookpro2159@CTRL-MacBook-Pro test % set COREHOST_TRACEFILE=text.txt
ctrl-macbookpro2159@CTRL-MacBook-Pro test % set COREHOST_TRACE=1
ctrl-macbookpro2159@CTRL-MacBook-Pro test % dotnet run
Hello World!
ctrl-macbookpro2159@CTRL-MacBook-Pro test % set COREHOST_TRACEFILE=text    
ctrl-macbookpro2159@CTRL-MacBook-Pro test % set COREHOST_TRACE=1       
ctrl-macbookpro2159@CTRL-MacBook-Pro test % dotnet run                 
Hello World!
ctrl-macbookpro2159@CTRL-MacBook-Pro test % ls 
Program.cs bin  obj  test..csproj
ctrl-macbookpro2159@CTRL-MacBook-Pro test %
```
Through the output results, I still did not see what we expected to see on the macOS system.

**Comment by hueifeng** [comment]
Take a look at the code snippet, the suffix in the folder is In the case of ` . `, the file cannot be found. 
If I modify the path and delete ` . `, the application executes successfully
```
ctrl-macbookpro2159@CTRL-MacBook-Pro test. % dotnet run 
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly '/Users/test./bin/Debug/netcoreapp3.1/test..dll'. The system cannot find the file specified.

File name: '/Users/test./bin/Debug/netcoreapp3.1/test..dll'

//path=/Users/test./bin/Debug/netcoreapp3.1/test..dll

ctrl-macbookpro2159@CTRL-MacBook-Pro test. % cd /Users/test 
ctrl-macbookpro2159@CTRL-MacBook-Pro test % dotnet run
Hello World!

//path=/Users/test/bin/Debug/netcoreapp3.1/test..dll
ctrl-macbookpro2159@CTRL-MacBook-Pro test %
```

**Comment by danmosemsft** [comment]
Well thanks for the report. Someone will take a look later. Meantime (unless you want to debug yourself) I suggest you get past it by not using the extra dot.

**Comment by hueifeng** [comment]
@danmosemsft  OK, Thank you.

**Comment by vitek-karas** [comment]
I don't have Mac at hand, but on Unix you have to set the environment variable via `export COREHOST_TRACE=1` otherwise it doesn't propagate to child processes.

**Comment by hueifeng** [comment]
@vitek-karas The results are as follows:
```
Launch host: /usr/local/share/dotnet/dotnet, app: /usr/local/share/dotnet/sdk/3.1.200/dotnet.dll, argc: 1, args: test..dll,
--- Begin breadcrumb write
Directory core breadcrumbs [] was not specified or found
Fallback directory core breadcrumbs at [opt/corebreadcrumbs] was not found
Breadcrumb store was not obtained... skipping write.
Could not execute because the specified command or file was not found.
Possible reasons for this include:
  * You misspelled a built-in dotnet command.
  * You intended to execute a .NET Core program, but dotnet-test..dll does not exist.
  * You intended to run a global tool, but a dotnet-prefixed executable with this name could not be found on the PATH.
Execute managed assembly exit code: 0x1

[进程已完成]

```

**Comment by vitek-karas** [comment]
I was able to repro this on Linux. Repro is as above, create a directory which ends with `.` and then `dotnet new console` and `dotnet run`.

The error comes from actually running the program from within managed code.

To get the correct traces one has to go to the output directory and run the executable directly. That will produce the host trace which is interesting, but the traces don't show anything broken. The assembly which fails is correctly added to the TPA.

**Comment by hueifeng** [comment]
It seems that `.` was removed during the dos path conversion.

https://github.com/dotnet/runtime/blob/578fe57d9068d0d697bf1680f0c6a2dde4cfceca/src/coreclr/src/pal/src/file/path.cpp#L549

**Comment by vitek-karas** [comment]
We should at least understand the actual problematic piece of code to decide if this is something to fix in .NET 5.

**Comment by KevinCathcart** [comment]
This was fixed for 5.0 in #459.

As far as I can tell, the only actionable item here is possibly backporting the fix, if deemed safe enough.

See also: #4093, #9648 which are both older duplicates of this same issue.

**Comment by danmosemsft** [comment]
Oh !  I should have seen that duplicate earlier!. @hueifeng can you confirm it's fixed in recent 5.0 preview?

**Comment by hueifeng** [comment]
@danmosemsft awesome! I tested it and found that it was fixed in the 5.0 preview.

**Comment by danmosemsft** [comment]
Thanks for verifying @hueifeng
